src/main/java/org/kairosdb/datastore/cassandra/CassandraDatastore.java

```diff
@@ -250,121 +240,65 @@ public void putDataPoint(String metricName,
-                            countsByMetricName = new ArrayList<>();
-                            countsByMetricName.addAll(insertCountByMetric.entrySet());
-                            insertCountByMetric.clear();
-                            logger.warn("RowKeys inserted: count={}", count);
-                        }
-                    }
-                }
-            }
-
-            if (writeRowKey) {
-
-                BoundStatement bs = new BoundStatement(m_psInsertRowKey);
-                bs.setBytes(0, ByteBuffer.wrap(metricName.getBytes(UTF_8)));
-
-                ByteBuffer serializedKey = DATA_POINTS_ROW_KEY_SERIALIZER.toByteBuffer(rowKey);
-
-                bs.setBytes(1, serializedKey);
-                bs.setInt(2, rowKeyTtl);
-                m_session.executeAsync(bs);
-
-                for (String split : m_indexTagList) {
-                    String v = tags.get(split);
-                    if (null == v || "".equals(v)) {
-                        continue;
-                    }
-
-                    bs = new BoundStatement(m_psInsertRowKeySplit);
-                    bs.setString(0, metricName);
-                    bs.setString(1, split);
-                    bs.setString(2, v);
-                    bs.setBytes(3, serializedKey);
-                    bs.setInt(4, rowKeyTtl);
-
-                    m_session.executeAsync(bs);
+            boolean writeRowKeyReverseLookups = false;
+            final boolean rowKeyKnown = rowKeyCache.isKnown(serializedKey);
+            if (!rowKeyKnown) {
+                writeRowKeyReverseLookups = true;
+                rowKeyCache.put(serializedKey);
+
+                int count = ++keyInsertCount;
+                if (insertCountByMetric.containsKey(metricName)) {
+                    insertCountByMetric.put(metricName, insertCountByMetric.get(metricName) + 1);
+                } else {
+                    insertCountByMetric.put(metricName, 1);
                 }
 
-                for (RowKeyListener rowKeyListener : m_rowKeyListeners) {
-                    rowKeyListener.addRowKey(metricName, rowKey, rowKeyTtl);
+                if ((writeTime - keyInsertTimer) > 30_000) {
+                    keyInsertTimer = writeTime;
+                    keyInsertCount = 0;
+                    countsByMetricName = new ArrayList<>(insertCountByMetric.entrySet());
+                    insertCountByMetric.clear();
+                    logger.warn("RowKeys inserted: count={}", count);
                 }
+            }
 
-                if (countsByMetricName != null) {
-                    Collections.sort(countsByMetricName, (x,y) -> y.getValue() - x.getValue());
-                    logger.warn("Keys inserted: {}", countsByMetricName.subList(0, Math.min(10, countsByMetricName.size())));
-                }
+            if (writeRowKeyReverseLookups) {
+                storeRowKeyReverseLookups(metricName, serializedKey, rowKeyTtl, tags, countsByMetricName);
```

Comment:
just put this statement inside previous `if` block and remove `writeRowKeyReverseLookups` variable. This block executes only if previous `if` block is executed